### PR TITLE
Avoid panic when reporting error in conformance test

### LINKED ISSUE #89769: Panic in conformance test (1.17.3)
Found in local testing. Looks like it's coming from:
https://github.com/kubernetes/kubernetes/blob/v1.17.3/test/utils/deployment.go#L201

```
STEP: Destroying namespace "webhook-7543-markers" for this suite.
[AfterEach] [sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/webhook.go:102

•! Panic in Spec Setup (BeforeEach) [17.889 seconds]
[sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin]
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/framework.go:23
  should deny crd creation [Conformance] [BeforeEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:721

  Test Panicked
  runtime error: index out of range [0] with length 0
  /usr/local/go/src/runtime/panic.go:75

  Full Stack Trace
  k8s.io/kubernetes/test/utils.WaitForDeploymentRevisionAndImage(0x5477b20, 0xc001f62580, 0xc0056fe6f0, 0xc, 0x4ae2953, 0x19, 0x4a894a8, 0x1, 0xc001ea0300, 0x2d, ...)
  	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/utils/deployment.go:201 +0x669
  k8s.io/kubernetes/test/e2e/framework/deployment.WaitForDeploymentRevisionAndImage(...)
  	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/deployment/wait.go:53
  k8s.io/kubernetes/test/e2e/apimachinery.deployWebhookAndService(0xc000c34500, 0xc001ea0300, 0x2d, 0xc00276a230, 0x20fc000020fb)
  	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/webhook.go:845 +0xba5
  k8s.io/kubernetes/test/e2e/apimachinery.glob..func19.1()
  	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/webhook.go:99 +0x2f7
  k8s.io/kubernetes/test/e2e.RunE2ETests(0xc0024f4300)
  	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:110 +0x30a
  k8s.io/kubernetes/test/e2e.TestE2E(0xc0024f4300)
  	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:112 +0x2b
  testing.tRunner(0xc0024f4300, 0x4c77cd8)
  	/usr/local/go/src/testing/testing.go:909 +0xc9
  created by testing.(*T).Run
  	/usr/local/go/src/testing/testing.go:960 +0x350
```

### PR DESCRIPTION
**Pull request by dims**
Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is for just making sure we report the correct error. 

**Which issue(s) this PR fixes**:

Fixes #89769

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/utils/deployment.go [hunk 1]
```diff
@@ -199,7 +199,15 @@ func WaitForDeploymentRevisionAndImage(c clientset.Interface, ns, deploymentName
 		return fmt.Errorf("deployment %q failed to create new replica set", deploymentName)
 	}
 	if err != nil {
-		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deployment.Spec.Template.Spec.Containers[0].Image, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRS.Spec.Template.Spec.Containers[0].Image, revision, image, err)
+		deploymentImage := ""
+		if len(deployment.Spec.Template.Spec.Containers) > 0 {
```
Comment (liggitt) [inline test/utils/deployment.go:203]: can't deployment be nil?

Comment (dims) [reply test/utils/deployment.go:203]: i can add that check as well.

File: test/utils/deployment.go [hunk 2]
```diff
@@ -199,7 +199,15 @@ func WaitForDeploymentRevisionAndImage(c clientset.Interface, ns, deploymentName
...
-		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deployment.Spec.Template.Spec.Containers[0].Image, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRS.Spec.Template.Spec.Containers[0].Image, revision, image, err)
+		deploymentImage := ""
+		if deployment != nil && len(deployment.Spec.Template.Spec.Containers) > 0 {
+			deploymentImage = deployment.Spec.Template.Spec.Containers[0].Image
+		}
+		newRSImage := ""
+		if newRS != nil && len(newRS.Spec.Template.Spec.Containers) > 0 {
+			newRSImage = newRS.Spec.Template.Spec.Containers[0].Image
+		}
+		return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deploymentImage, newRS.Name, newRS.Annotations[deploymentutil.RevisionAnnotation], newRSImage, revision, image, err)
```
Comment (neolit123) [inline test/utils/deployment.go:210]: the deployment == nil check should be before the err != nil

```
...
if newRS == nil {
	return fmt.Errorf("deployment %q failed to create new replica set", deploymentName)
}
if deployment == nil {
	return fmt.Errorf("could not get deployment %q", deploymentName)
}
if err != nil {
...
```

another improvement would be to expose the underlying error from the poll.

Comment (liggitt) [reply test/utils/deployment.go:210]: newRS.Name is still accessed directly and could NPE

Comment (dims) [reply test/utils/deployment.go:214]: fixed.

Comment (liggitt) [reply test/utils/deployment.go:214]: are you sure? I still see it here:

return fmt.Errorf("error waiting for deployment %q (got %s / %s) and new replica set %q (got %s / %s) revision and image to match expectation (expected %s / %s): %v", deploymentName, deployment.Annotations[deploymentutil.RevisionAnnotation], deploymentImage, **newRS.Name**, newRS.Annotations[deploymentutil.RevisionAnnotation], newRSImage, revision, image, err)

Comment (dims) [reply test/utils/deployment.go:214]: @liggitt we bail out in line 199 i think `newRS == nil`

Comment (liggitt) [reply test/utils/deployment.go:214]: ah, got it

**Pull request merged**